### PR TITLE
Fix MLX Metal resource-limit crash in finite text training via CPU batch plans and bounded compile signatures

### PR DESCRIPTION
**Pull request by Lyxot**
Partially addresses unslothai/unsloth#6990 (do not auto-close): this PR bounds the Metal resource growth for the finite text training path only; VLM training can still accumulate resources the same way and will be covered by a follow-up extending the same plan/guard architecture, so the issue should stay open after this merges.

## Summary

Finite MLX text training could exhaust Metal's private allocator resource limit (`[metal::malloc] Resource limit (499000) exceeded`) on datasets with many distinct sequence lengths: on a real reproduction over the 1,985 distinct FineTome-100k widths at a 2,048-token cap, main fails after 1,010 steps at a 498,920-resource peak. This PR removes the three application-controlled sources of that growth for finite, non-VLM text training: the live `O(microsteps)` list of pre-materialized MLX input arrays, the unbounded set of distinct `(batch schema, sequence width)` signatures presented to `mx.compile`, and runtime CCE's unbounded chunk-plan cache. The same reproduction now completes all 1,985 steps with 32 compile signatures and a 29,967-resource peak (94.0% below main's highest successful sample), and bucketed workloads train 24-29% faster.

## Root cause

The finite text trainer pre-materialized every scheduled batch as MLX arrays up front and kept the whole list alive for the run, and each distinct padded width formed a new compiled-callable signature, so allocator resources grew with the number of distinct widths in the dataset rather than with any bounded quantity. Runtime CCE added one cached chunk-plan array set per new logits geometry on top of that. None of these were bounded, so long or width-diverse schedules eventually hit the hard Metal resource ceiling mid-training.

## What changed

- Batch storage: finite text schedules become a CPU-backed batch plan (row token data plus a schedule of index groups); MLX arrays are materialized per visit instead of being pre-built and retained, so live scheduled inputs on Metal are bounded by the current batch only.
- Compile-signature planning: a deterministic bounded planner assigns each batch a padded width endpoint so the complete set of compiled signatures is capped (algorithm below); materialization pads upward-only to the planned endpoint and admission-checks every `(schema family, raw width, accumulation phase)` against the planned catalog before any compiled call, with the existing eager fallback (or strict-mode error) when a schedule cannot be planned.
- DDP: ranks coordinate failure/mode and synchronize the maximum requested automatic cap before model mutation, then each rank materializes its deterministic snapshot under that shared cap.
- Runtime CCE: chunk-plan arrays are canonicalized and bounded instead of growing per logits geometry.
- Deterministic visits: eligible one-cycle schedules now expose a pure `batch_index_for_visit(absolute_visit)` mapping — epoch 0 preserves the stored order and later epochs visit a deterministic plan-local permutation of the same batch multiset — consumed identically by the shape-guard enumeration, the runtime fetch, both compile-failure retry paths, and checkpoint-resume fast-forward, so the guard's planned catalog always equals the visited signature stream.
- Width policy: the three equivalent pad-width computations (plan `batch_width`, response-masked explicit widths, streaming batch builder) are consolidated into one shared helper with a fixed round-up → minimum → cap order and per-caller policy preserved.

## Planning algorithm

A complete text signature is `(compile scope, gradient-accumulation phase, batch schema family, padded width endpoint)`, where the family is the compile-relevant batch schema excluding sequence width. The planner enumerates the schedule's `(family, raw width, phase, frequency)` events and then:

- Schedules with at most 32 full signatures stay exact and unpadded (verified zero added padding on such schedules).
- Larger automatic schedules build one deterministic bounded frontier by repeatedly merging adjacent width buckets within each family (padding is upward-only, so merges only stretch narrower batches to the bucket endpoint), retaining candidate snapshots up to a 128-signature ceiling. The selected plan is the smallest-signature snapshot that satisfies both exact budgets — at most 5% added quadratic attention work and at most 1.5x per-batch width stretch — labeled `padding_budget`; if no snapshot satisfies both, the minimum-padding snapshot under the ceiling is used and labeled `ceiling`.
- An explicit integer `compile_max_variants` from 1 to 256 is authoritative (`fixed`): within-cap schedules stay exact and larger ones use the same bounded frontier under the requested cap.
- The labels describe policy outcomes, not global optimality, and the planner guarantees only application-side invariants (bounded signatures, bounded live batches, bounded CCE plans); it does not attempt to estimate Metal's internal resource count.

## Behavior changes reviewers should know about

- Deliberate seed canonicalization: schedule construction now uses a plan-local RNG through seed normalization, so `seed=None` (previously ambient process RNG), the default creator's `seed=0` path (previously ambient), and response-masked `seed=None` ordering (previously entropy-derived) become deterministic; explicitly seeded schedules keep byte-identical output because the plan-local generator reproduces the former global-seed stream. This is required so fresh-process checkpoint resume can reconstruct the schedule.
- Deliberate epoch semantics change: eligible one-cycle schedules no longer replay an identical batch order every epoch; later epochs are deterministic permutations of the same batch multiset. Horizon-expanded, truncated, ordered, response-masked, and directly constructed plans keep identity visits via explicit construction metadata.
- Padding changes tensor shapes only: padded positions carry the pad id, attention 0, and label -100, and loss stays token-count-normalized, so metrics are unchanged; the full-schedule reproduction adds 3.47% padded tokens for a 4.91% work increase within budget.

## Performance

Benchmarked on Apple M3 Max, MLX 0.32.0, fresh process per training cell, no MLX memory/cache limits, strict full-step compile with CCE and gradient checkpointing; synthetic and FineTome runs use cached 4-bit Gemma 3 270M QLoRA rank 4 batch 1, the notebook block uses the canonical Gemma3 270M notebook configuration (native precision, rank-128 LoRA, batch 4, response-only, 100 steps). Counterbalanced main/automatic/automatic/main ordering with matching dataset and schedule digests.

Planner-only cost and plan quality (fresh process each; the small schedule takes the production no-padding bypass):

| Schedule | Raw signatures | Planned signatures | Median planning time | Added quadratic work | Max stretch |
|---|---:|---:|---:|---:|---:|
| Small | 24 | 24 | 0.032 ms | 0.00% | 1.000x |
| Notebook | 35 | 10 | 0.990 ms | 4.52% | 1.285x |
| Clustered | 64 | 8 | 3.350 ms | 4.97% | 1.500x |
| Irregular | 96 | 16 | 7.660 ms | 4.91% | 1.494x |
| Long-tail | 144 | 22 | 16.421 ms | 4.69% | 1.344x |
| Cached medium schedule | 600 | 27 | 61.893 ms | 4.95% | 1.346x |
| Full FineTome | 1,985 | 32 | 209.250 ms | 4.91% | 1.472x |

Synthetic end-to-end training (two-run means):

| Schedule | Time vs main | Sampled resources vs main | Added work |
|---|---:|---:|---:|
| Small | -8.6% (noise; no bucketing, no padding) | -5.2% | 0.00% |
| Clustered | -24.9% | -72.2% | 4.97% |
| Irregular | -23.6% | -70.7% | 4.91% |
| Long-tail | -29.2% | -73.9% | 4.69% |

Canonical Gemma 3 270M notebook: 35 → 10 signatures and 21,659 → 12,043 peak resources (-44.4%); mean time 246.54 s → 224.34 s, but pass-to-pass variance under desktop load supports claiming performance parity rather than a guaranteed speedup.

600-width medium control: 600 → 34 signatures (-94.3%), 325,751 → 32,166 mean peak resources (-90.1%), 578.65 s → 212.76 s mean; peak RSS 7.376 GiB → 1.797 GiB. Full 1,985-width reproduction: main fails at step 1,010 of 1,985 at 498,920 resources; automatic completes all steps at 3,410.5 useful tok/s with 32 signatures and a 29,967-resource peak, both padding budgets satisfied and losses finite.

## Risks and tradeoffs

- Bucketed schedules pay bounded extra attention work (at most 5% quadratic work, at most 1.5x per-batch stretch by construction; 3.47% padded tokens on the full reproduction), in exchange for far fewer compilations and bounded allocator growth.
- Planning cost is one-time and small relative to training (209 ms for 1,985 distinct widths).
- Guarantees are application-side invariants only; the planner does not model Metal's internal accounting, and unsupported modes (VLM, streaming, preference training) keep their existing behavior with the guard reporting not-applicable.
- The seed-canonicalization and epoch-order changes above are intentional behavior changes for previously nondeterministic or replay-only cases; explicitly seeded schedules are byte-identical.

## Suggested review path

The history intentionally keeps the development sequence; some middle commits introduce planner mechanics that later commits simplify, so reviewing by group against the final state is more efficient than strict chronology:

- CPU finite-plan substrate: `refactor(mlx): separate finite text batch plans`, `refactor(mlx): materialize finite text batches lazily`.
- Guard, DDP, and CCE contracts: `feat(mlx): optimize bounded text shape buckets` through `fix(mlx): bound runtime CCE chunk plans` (review the group's net effect by final file).
- Cap and padding-budget policy: `policy(mlx): raise finite text compile cap` through `fix(mlx): retain stretch-feasible text buckets`.
- Planner simplification arc: `perf(mlx): bound large text shape planning` through `refactor(mlx): use bounded text shape planning` (net effect removes exact dynamic programming; -632 net lines).
- Semantics-bearing follow-ups: `feat(mlx): deterministic absolute visits for finite text plans` (data-order and seed canonicalization changes described above), `refactor(mlx): consolidate finite text pad-width arithmetic` (arithmetic-only).

## Validation

```bash
python -m pytest tests/test_mlx_shape_guard.py tests/test_mlx_batch_padding.py tests/test_mlx_batching_and_decay.py tests/test_mlx_trainer_internals.py -q
# 158 passed, 5 skipped
python -m pytest tests/test_pr684_review_fixes_a.py -q -k "not thread4"
# 12 passed (thread4 SGD-vs-torch comparisons fail identically on main in this environment)
python -m compileall -q unsloth_zoo/mlx/utils.py unsloth_zoo/mlx/trainer.py && git diff --check
```

Development ran through repeated multi-reviewer rounds per stage plus the counterbalanced benchmark campaign above; the deterministic-visits and width-consolidation commits each went through three to four review rounds with independent reviewers before landing.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -3371,6 +3373,326 @@ def _labeled_row_has_supervision(labels, max_seq_length):
...
+            for row_index in batch_indices
+        ]
+        if self._widths is not None:
+            return self._widths[int(index)]
+        return _finite_text_pad_width(
+            max(lengths, default=0),
+            pad_to_multiple=self.pad_to_multiple,
+            minimum_width=self.minimum_width,
+            max_seq_length=self.max_seq_length,
+        )
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py:3541]: Fixed in 6a85a0aa: the explicit-width return now precedes the length scan, so explicit-width (response-masked) plans skip the per-row work entirely; IndexError semantics are unchanged because the constructor validates widths/schedule alignment. The second half of the suggestion (having materialize() hand its own lengths/raw_width back into batch_width) is deliberately not taken: it would widen the signature or duplicate the width policy at a second call site, while batch_width stays the single stateless authority the guard preflight and runtime both rely on, and the remaining scan is O(local batch size) on the non-explicit path only.

File: unsloth_zoo/mlx/trainer.py [hunk 2]
```diff
@@ -664,6 +682,161 @@ def __init__(self, *args, **kwargs):
...
+        batch_index = batches.batch_index_for_visit(microstep)
+        family = batches.batch_family(batch_index)
+        width = batches.batch_width(batch_index)
+        phase = phase_for_microstep(
+            compile_scope,
+            args.gradient_accumulation_steps,
+            microstep,
+        )
+        key = (family, width, phase, len(batches.schedule[batch_index]))
+        event_counts[key] = event_counts.get(key, 0) + 1
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py:781]: Leaving this as-is intentionally. The enumeration is a one-time preflight, and its cost is included in the PR's planner-matrix medians (0.03 ms at 24 signatures up to 209 ms at 1,985 distinct widths). A closed-form aggregation would still have to simulate the per-epoch visit permutations to pair widths with accumulation phases correctly, so it saves roughly the grad-accum constant factor while reintroducing the analytic planning complexity this branch's simplification arc deliberately removed. If profiling ever shows preflight cost mattering on real runs we can revisit with that evidence.

File: tests/test_mlx_ddp_metal.py [hunk 3]
```diff
@@ -393,7 +394,7 @@ def make_vlm_compile_parity_trainer(output_name, compile_enabled):
     cmd = [
         str(launcher), "-n", "2", "--backend", "ring",
         "--python", sys.executable, "--cwd", str(repo_root),
-        str(script), str(tmp_path),
+        sys.executable, str(script), str(tmp_path),
     ]
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply tests/test_mlx_ddp_metal.py:398]: Left as-is because the launcher does not double-invoke Python: mlx.launch execs the positional argv verbatim (its launch script builds cmd=(<command>); exec "${cmd[@]}"), and --python selects the interpreter only for remote-host bootstrap (help text: 'Use this python on the remote hosts'). Locally this therefore runs exactly '<python> script.py <tmp_path>' with no duplicated interpreter, and keeping both the flag and the positional interpreter makes the test correct on localhost and over ssh hostfiles alike. The DDP Metal suite passes as written.

**Comment by danielhanchen** [comment]
Pushed a fix on top of this.

The compiled full step threads `mx.random.state` through its `outputs`, so a trace-time `mx.compile` failure can leave the global RNG state partially updated. The DDP local-grad path already snapshots and restores it before the eager retry, but the single-process path did not, so a runtime compile fallback resumed from a corrupted RNG state (I reproduced a crash on the eager retry). The commit mirrors the DDP snapshot/restore in the single-process branch, guarded on the list form so the torch-sim test shim stays a no-op. It also drops the now-unused `create_ordered_batches` import. The MLX suite passes per file on CPU, and the new tests plus the pr684 validation pass on an Apple Silicon runner.

A few non-blocking notes for follow-up:

- The "Risks and tradeoffs" section states the plan pays "at most 5% quadratic work, at most 1.5x per-batch stretch by construction". That holds for `padding_budget` plans, but a `ceiling` plan (chosen when no snapshot under the 128 signature cap meets both budgets) stays signature-bounded and can exceed either budget. The "Planning algorithm" section already describes this accurately, so it is worth softening the "Risks" wording to match.
- Epoch-based finite runs (`num_train_epochs > 0`, `max_steps <= 0`) now reshuffle each epoch instead of replaying the epoch 0 order. Determinism is preserved since the permutation is a function of `(seed, epoch)`, but this changes loss curves across the upgrade and an old checkpoint resumed past epoch 0 will visit a different order. Worth a changelog line.
- Evaluation still builds eager batches via `create_batches()` for both ordinary and response-masked splits, so a large or width-diverse eval set can still grow allocator resources. Fine as a follow-up given this PR is scoped to training.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 4]
```diff
@@ -4611,13 +4948,45 @@ def __len__(self):
...
+        [dataset[index] for index in range(len(dataset))],
+        with_offsets=True,
+    )
+    return FiniteTextBatchPlan(
+        rows,
+        schedule,
+        max_seq_length=max_seq_length,
+        pad_id=0,
+        visit_policy="epoch_permute" if num_batches is None else "identity",
+        visit_seed=seed,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:4982]: Fixed in 28f869a1: the default text plan now pads to mlx-lm's 1 + 32*ceil(len/32) width (minimum_width=2, pad_to_multiple=32), matching the distributed and streaming builders. Added a regression test.

**Comment by danielhanchen** [comment]
@codex review
